[jit] bind autograd.backward and tensor.backward in TorchScript

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23913 [jit] bind autograd.backward and tensor.backward in TorchScript**

Summary:
This PR bind torch.autograd.backward and tensor.backward to TorchScript,
and make aliasing to the conservative for these two ops, this is mainly
because backward op might write to every input tensors in the graph

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23913

Differential Revision: [D16923272](https://our.internmc.facebook.com/intern/diff/D16923272)

zdevito commented on torch/csrc/jit/passes/alias_analysis.cpp:
```diff
@@ -665,6 +667,35 @@ void AliasDb::analyzeWait(Node* node) {
   }
 }
 
+void AliasDb::analyzeBackward(Node* node) {
```
This function doesn't handle a ton of aliasing cases. For instance, what if the grad tensor is stored in a list inside a class somewhere. A better way to do this would be to mark backward as CONSERVATIVE.

zdevito commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -612,6 +621,18 @@ RegisterOperators reg(
+           self.backward(gradient, keep_graph, create_graph);
+           return 0;
+         },
+         aliasAnalysisSpecialCase()),
```
Just put aliasAnalysisConservative here and this will be correct.

wanchaol replied:
The alias_analaysis does not allow me to make it a conservative kind because of the assertion [here](https://github.com/pytorch/pytorch/blob/c6617b370b903b74a5647dd34e69fd0081713df8/torch/csrc/jit/passes/alias_analysis.cpp#L395). So what I did is that I made them the special_case kind and call `analyze_conservative` when handling these special cases, this happens before the assert so it should work properly.

zdevito replied:
That assertion is bogus, just remove it.

zdevito commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -49,7 +49,16 @@ namespace jit {
 namespace {
 
-template<class T>
+// XXX: This function is to specialize IValue for tensor type in
```
This is a copy/paste that propagates a comment that is incorrect in this context. Let's just inline this instead:

```
auto gradient_ivalue = pop(stack);
auto gradient = gradient_ivalue.isNone() ? at::Tensor() : gradient_ivalue.toTensor()
```

zdevito reviewed (commented):
Alias analysis changes have some issues.  There should be at least one test that ensures that optimization of grad tensors knows about the mutation of grad tensors.

wanchaol committed:
Update on "[jit] bind backward in TorchScript"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23913

wanchaol committed:
Update on "[jit] bind backward in TorchScript"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23913

wanchaol committed:
Update on "[jit] bind backward in TorchScript"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23913

wanchaol committed:
Update on "[jit] bind backward in TorchScript"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23913

wanchaol committed:
Update on "[jit] bind backward in TorchScript"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23913

wanchaol committed:
Update on "[jit] bind autograd.backward and tensor.backward in TorchScript"


Summary:
This PR bind torch.autograd.backward and tensor.backward to TorchScript,
and make aliasing to the conservative for these two ops, this is mainly
because backward op might write to every input tensors in the graph

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23913

wanchaol committed:
Update on "[jit] bind autograd.backward and tensor.backward in TorchScript"


Summary:
This PR bind torch.autograd.backward and tensor.backward to TorchScript,
and make aliasing to the conservative for these two ops, this is mainly
because backward op might write to every input tensors in the graph

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23913

wanchaol committed:
Update on "[jit] bind autograd.backward and tensor.backward in TorchScript"


Summary:
This PR bind torch.autograd.backward and tensor.backward to TorchScript,
and make aliasing to the conservative for these two ops, this is mainly
because backward op might write to every input tensors in the graph

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23913

Differential Revision: [D16923272](https://our.internmc.facebook.com/intern/diff/D16923272)

wanchaol committed:
address comments on "[jit] bind autograd.backward and tensor.backward in TorchScript"


Summary:
This PR bind torch.autograd.backward and tensor.backward to TorchScript,
and make aliasing to the conservative for these two ops, this is mainly
because backward op might write to every input tensors in the graph

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23913

Differential Revision: [D16923272](https://our.internmc.facebook.com/intern/diff/D16923272)

zdevito reviewed (approved):
This is good, but let's fix aliasAnalysisConservative rather than special casing backward.

wanchaol committed:
open up conservative kind aliasing for op registration on "[jit] bind autograd.backward and tensor.backward in TorchScript"


Summary:
This PR bind torch.autograd.backward and tensor.backward to TorchScript,
and make aliasing to the conservative for these two ops, this is mainly
because backward op might write to every input tensors in the graph

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23913

Differential Revision: [D16923272](https://our.internmc.facebook.com/intern/diff/D16923272)

wanchaol committed:
Update on "[jit] bind autograd.backward and tensor.backward in TorchScript"


Summary:
This PR bind torch.autograd.backward and tensor.backward to TorchScript,
and make aliasing to the conservative for these two ops, this is mainly
because backward op might write to every input tensors in the graph

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23913

Differential Revision: [D16923272](https://our.internmc.facebook.com/intern/diff/D16923272)

wanchaol committed:
rebase on "[jit] bind autograd.backward and tensor.backward in TorchScript"


Summary:
This PR bind torch.autograd.backward and tensor.backward to TorchScript,
and make aliasing to the conservative for these two ops, this is mainly
because backward op might write to every input tensors in the graph

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23913

Differential Revision: [D16923272](https://our.internmc.facebook.com/intern/diff/D16923272)
